### ISSUE TITLE
/oidc/callback/ infinite redirect loop

### ISSUE THREAD

**Issue by willkg** [mozilla/mozilla-django-oidc#153]
I switched standups from our home-grown auth0 client implementation to mozilla-django-oidc 0.3.1 on Friday. At least one user went to use the site and immediately entered an infinite redirect on this url:

```
https://www.standu.ps/oidc/callback/?error=login_required&error_description=Login%20required&state=EYy2Qz8peNMCImxor3RKHGAGf8ggKLvh
```

Clearing the browser's standup cookies fixed it.

It's possible that it's the `/oidc/callback/` view does its thing and there's one branch that redirects to itself or something. Maybe this is a problem in conjunction with the middleware?

**Comment by willkg** [comment]
I have a few users who keep hitting the infinite redirect, so I don't think the theory that it was a transition issue holds water.

Maybe what's going on is that the browser goes to `/oidc/callback` with an error. The callback view has no explicit handling for this. The callback view redirects the user to the failure url which by default is `/`. Standup is using the middleware, so then the middleware processes that request and sends the browser back to Mozilla SSO which sends the browser back to `/oidc/callback` with the same error. Repeat.

That sure seems like what's going on.

Assuming that's true, I think to fix this we should:

1. add handling in `/oidc/callback` for when the querystring params have `error` and `error_description` and redirect to the failure url carrying that error code along
2. the failure url should not default to `/`--it should be an explicit view for "Yo--you fail at login." messages
3. the failure url should be exempted by the middleware

@johngian @akatsoulas @peterbe Does that sound right/likely?

**Comment by pmac** [comment]
Seeing this on basket as well.

**Comment by peterbe** [comment]
`error` and `error_description` aren't particular oidc specific so you couldn't do something like `if request.GET.get('error') and request.GET.get('error_description')` in the middleware. 

I too think the right thing to do is to have a another URL to redirect to that is not `/` and then the RP is responsible for making such a page. And the middleware should then do something like `if settings.RP_ERROR_URL == request.path: return`.

**Comment by peterbe** [comment]
This is likely to be very related: https://github.com/mozilla/airmozilla/issues/827#issuecomment-374274903
What happened there was that I was logged in (thank you django session cookies) but since my `oidc_id_token_expiration` has expired so mozilla_django_oidc redirects me to `/authorize?...&prompt=none` which then errors out and redirects back to `/oidc/callback` which redirects to `/login-failure/` (which is a public django page) and there the middleware kicks in and redirects back again. 

Above @willkg wrote this suggestion:

*"add handling in `/oidc/callback` for when the querystring params have `error` and `error_description` and redirect to the failure url carrying that error code along"*

I wonder if we should do something like this, in the `/oidc/callback`, instead: 
```python
if request.GET.get('error'):
    # Make sure the user is not logged in any more because the callback
    # was unhappy with the user. 
    if request.user.is_authenticated:  # XXX Do this the backwards compat way!
        auth.logout(request) 
```

Otherwise, if you allow the user to remain logged in on an old session cookie and Auth0 is grumpy the RefreshIDToken middleware will get stuck in a redirect loop.

**Comment by peterbe** [comment]
@johngian Please review this last comment. I think it's important. [Airmozilla](https://github.com/mozilla/airmozilla/issues/827) is currently busted at the moment and lots of people can't log in to view private videos without manually clearing their cookies.

**Comment by akatsoulas** [comment]
This should be resolved after #213.
